st2auth_ldap_backend/ldap_backend.py

```diff
@@ -71,6 +76,66 @@ def _scope_to_ldap_option(self, scope):
             opt = ldap.SCOPE_SUBTREE
         return opt
 
+    def _get_ldap_search_results(self, connection, username, criteria, result_id, current_ref_hop):
+        results = []
+        r_type, r_data = connection.result(result_id, all=0)
+
+        if r_type == ldap.RES_SEARCH_ENTRY:
+            results.append(self._get_ldap_search_entry(r_data))
+
+            _results = self._get_ldap_search_results(connection, username, criteria, result_id,
+                                                     current_ref_hop)
+            # This condition prevents to append empty results.
+            if _results:
+                results.append(_results)
+        elif r_type == ldap.RES_SEARCH_REFERENCE:
+            _results = self._get_ldap_search_referral(r_data, username, criteria, current_ref_hop)
+            if _results:
+                results.append(_results)
+
+            _results = self._get_ldap_search_results(connection, username, criteria, result_id,
+                                                     current_ref_hop)
```

Comment:
Can you provide some comments why a recursive call to _get_ldap_search_results is needed?